File: backend/src/main/java/com/sgs/sistema/entidade/Solicitacao.java
package com.sgs.sistema.entidade;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "solicitacao")
@Getter
@Setter
@NoArgsConstructor
public class Solicitacao {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "solicitante_id", nullable = false)
    private Solicitante solicitante;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "categoria_id", nullable = false)
    private Categoria categoria;

    @Column(columnDefinition = "TEXT")
    private String descricao;

    @Column(nullable = false, precision = 15, scale = 2)
    private BigDecimal valor;

    @Column(name = "data_solicitacao", nullable = false, updatable = false)
    private LocalDateTime dataSolicitacao;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private StatusSolicitacao status;

    // Regra de negócio: Toda solicitação deve iniciar com status SOLICITADO
    @PrePersist
    public void prePersist() {
        // Se a data não for informada, adiciona a atual
        if (this.dataSolicitacao == null) {
            this.dataSolicitacao = LocalDateTime.now();
        }      
        
        if (this.status == null) {
            this.status = StatusSolicitacao.SOLICITADO;
        }
    }
}

File: frontend/src/types/solicitacao.ts
export type Solicitacao = {
  id: number;
  solicitanteId: number;
  solicitanteNome: string;
  solicitanteDocumento: string;
  categoriaId: number | null;
  categoriaNome: string;
  descricao: string | null;
  valor: number;
  dataSolicitacao: string | null;
  status: StatusSolicitacao;
};

export type CreateSolicitacaoPayload = {
  solicitanteId: number;
  categoriaId: number;
  descricao: string;
  valor: number;
};

export type UpdateStatusPayload = {
  status: StatusSolicitacao;
};

export type StatusSolicitacao =
  | "SOLICITADO"
  | "LIBERADO"
  | "APROVADO"
  | "REJEITADO"
  | "CANCELADO";

export type ListarSolicitacoesParams = {
  status?: StatusSolicitacao;
  categoriaId?: number;
  dataInicio?: string;
  dataFim?: string;
};
